File: CPP/Algorithms/reverseStringWoSpecialChars.cpp
#include <iostream>
#include <ctype.h>
#include <string.h>

using namespace std;

void reverseStringWoSpecialChars(char *str)
{
    int startPtr = 0;
    int endPtr = strlen(str) - 1;

    while (startPtr < endPtr)
    {
        if (!isalpha(str[startPtr]))
            startPtr++;
        else if (!isalpha(str[endPtr]))
            endPtr--;
        else
        {
            swap(str[startPtr], str[endPtr]);
            startPtr++;
            endPtr;
        }
    }
}

// Test Code

int main()
{
    char str[] = "a!!!b.c.d,e'f,ghi";
    cout << "Original String : " << str << endl;
    reverseStringWoSpecialChars(str);
    cout << "Reversed without affecting special chars : " << str << endl;
    return 0;
}

File: Python/Algorithms/reverseStringWoSpecialChars.py
def reverseStringWoSpecialChars(string):
    string = list(string)

    startPtr = 0
    endPtr = len(string) - 1

    while (startPtr < endPtr):
        if not string[startPtr].isalpha():
            startPtr += 1
        elif not string[endPtr].isalpha():
            endPtr -= 1
        else:
            string[startPtr], string[endPtr] = string[endPtr], string[startPtr]
            startPtr += 1
            endPtr -= 1

    return "".join(string)

# Test Code


string = "a!!!b.c.d,e'f,ghi"
print("Original String : " + string)
rev_string = reverseStringWoSpecialChars(string)
print("Reversed String w/o affecting special chars : " + rev_string)
